// less-plugin-module-resolver/lib/alias-file-manager.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AliasFileManager = void 0;
const less_1 = __importDefault(require("less"));
const normalize_configs_1 = require("./normalize-configs");
class AliasFileManager extends less_1.default.FileManager {
    options;
    matchers;
    constructor(options) {
        super();
        this.options = options;
        this.matchers = (0, normalize_configs_1.getOptions)(options);
    }
    supports(filename) {
        return this.matchers.some(([matcher]) => matcher.test(filename));
    }
    supportsSync(filename) {
        return this.supports(filename);
    }
    loadFile(filename, currentDirectory, options, environment) {
        const newFilename = this.matchAndReplace(filename);
        return super.loadFile(newFilename, currentDirectory, options, environment);
    }
    loadFileSync(filename, currentDirectory, options, environment) {
        const newFilename = this.matchAndReplace(filename);
        return super.loadFileSync(newFilename, currentDirectory, options, environment);
    }
    matchAndReplace(filename) {
        let newFilename = filename;
        for (const [matcher, replace] of this.matchers) {
            if (matcher.test(newFilename)) {
                newFilename = replace(newFilename);
                break;
            }
        }
        return newFilename;
    }
}
exports.AliasFileManager = AliasFileManager;
